book/src/developers/designs/subentries.md

```diff
@@ -0,0 +1,131 @@
+# Sub-Entries
+
+As Kanidm has grown we have encountered issues with growing complexity of values and valuesets. These
+can be hard to create and add, they touch a lot of the codebase, and they add complexity to new
+features or changes.
+
+These complex valueset types (such as authsession, oauth2session, application passwords) arose out
+of a need to have data associated to an account, but that data required structure and nesting
+of certain components.
+
+Rather than continue to add more complex and unwieldy valuesets, we need a way to create entries
+that refer to others.
+
+## Existing Referential Code
+
+The existing referential integrity code is designed to ensure that values from one entry are removed
+cleanly if the referenced entry is deleted. As an example, a group with a member "ellie" should have
+the reference deleted when the entry "ellie" is deleted.
+
+If the group were deleted, this has no impact on ellie, since the reference is defining a weak
+relationship - the user is a member of a group.
+
+## What Is Required
+
+What we need in a new reference type are the following properties.
```

Comment:
A lot of this section should be simply handled with foreign keys in the database, we shouldn't reinvent the wheel on this one.

Reply:
@yaleman We can't, because we have replication and replication conflicts. *We* are the database and *we* are implementing foreign keys here, so we need to be aware of things.